### PR TITLE
feat(optimizer): Multi join reordering

### PR DESCRIPTION
**Pull request by jon-chuang**
## What's changed and what's your intention?
Add a simple join reordering heuristic that attempts to perform left-deep join of connected components so that adjacent joins always have a eq condition edge between them.

Future work is to make the join ordering heuristic more sophisticated.

Enables e2e tests of TPCH q4, q8, q9 without needing manual join reordering to avoid cross joins.

**Replaced with:** https://github.com/singularity-data/risingwave/pull/2651 (needed to run CI)

Sequel to: https://github.com/singularity-data/risingwave/pull/2526
Tracking issue: https://github.com/singularity-data/risingwave/issues/2530
Fixes: https://github.com/singularity-data/risingwave/issues/1866

Issues:
- This reorders a lot of joins, so it causes a lot of planner output tests to fail. The planner tests are somewhat brittle as there is currently no canonical ordering (the ordering is random as long as the join tree is connected). Not sure what the right thing to do here is. 
  - Would be nice to push the projection back to the top of the query plan as it is a simple projection, thus removing it (don't know what is the runtime cost of simple projection)

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 1]
```diff
@@ -111,6 +115,269 @@ impl PlanTreeNode for LogicalMultiJoin {
...
+                )
+                .into()
+            });
+
+        if join_ordering != (0..self.input_col_nums().iter().sum()).collect::<Vec<_>>() {
+            output =
+                LogicalProject::with_mapping(output, self.mapping_from_ordering(join_ordering));
+        }
+
+        // We will later push down the `non_eq_cond` back to the individual joins via the
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:249]: not just non_eq_cond, all conditions including join conditions

**Comment by st1page** [comment]
> This reorders a lot of joins, so it causes a lot of planner output tests to fail

the planner test cases are consider as a regression test, so I think just `./risedev do-apply-planner-test` and we can see what changes on the plan and check if it is ok.

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 2]
```diff
@@ -111,6 +115,269 @@ impl PlanTreeNode for LogicalMultiJoin {
...
+            return;
+        }
+
+        // Reassign to the smaller label
+        let old_vertices = self.labels_to_vertices.remove(&old_label).unwrap();
+        self.labels_to_vertices
+            .get_mut(&new_label)
+            .unwrap()
+            .extend(old_vertices.iter());
+        for v in old_vertices {
```
Comment (lmatz) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:172]: Seems `Union-Find` can help find the connected components more efficiently. 

But the number of joins won't be too many, the current implementation should also be ok.

**Comment by st1page** [comment]
> Would be nice to push the projection back to the top of the query plan as it is a simple projection, thus removing it (don't know what is the runtime cost of simple projection)

-1, the simple projection will prune the unnecessary columns, and the unnecessary columns will be an expensive overhead in some executor such as HashJoin.

**Comment by skyzh** [comment]
> This reorders a lot of joins, so it causes a lot of planner output tests to fail. The planner tests are somewhat brittle as there is currently no canonical ordering (the ordering is random as long as the join tree is connected). Not sure what the right thing to do here is.

We can use deterministic structures for storing data. e.g., BTreeMap instead of HashMap. Therefore, the result shall be stable.

**Comment by jon-chuang** [comment]
I've decided to sort by lexicographic ordering until we have the heuristic implemented.

**Comment by jon-chuang** [comment]
> ./risedev do-apply-planner-test and we can see what changes on the plan and check if it is ok.

Ok, but I think its most of the planner output since many tests have a join. So its quite hard to verify by inspection. 

At least all the e2e tests pass.

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs (hunk 2 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_multi_join.rs:172]: yeah its definitely not best performance. Didn't want to write a custom data structure just for it...

**Comment by jon-chuang** [comment]
closed in favour of: https://github.com/singularity-data/risingwave/pull/2651

**Pull request closed without merging**